File: ui/src/lib/types/pkg.ts
import type { User } from "./user";

export type ProjectVisibility = "Public" | "Private" | "Unlisted";

export interface Package {
    /**
     * The ID of the package.
     */
    id: number;

    /**
     * The package name.
     */
    name: string;

    /**
     * The slug (URL identifier) of the package.
     */
    slug: string;

    /**
     * The content of this package's README file.
     */
    readme: string;

    /**
     * This package's description.
     */
    description: string;

    /**
     * An optional link to this package's source code.
     */
    source?: string;

    /**
     * An optional link to this package's issue tracker.
     */
    issues?: string;

    /**
     * An optional link to this package's wiki.
     */
    wiki?: string;

    /**
     * A stringified {@link Date} for when this version was created.
     */
    created_at: string;

    /**
     * A stringified {@link Date} for when this version was updated.
     */
    updated_at: string;

    /**
     * The amount of views this package gets.
     */
    views: number;

    /**
     * The visibility of a package.
     */
    visibility: ProjectVisibility;

    /**
     * The license a package is under.
     */
    license?: string;
}

export interface PackageUpdate {
    /**
     * The package name.
     */
    name?: string;

    /**
     * The content of this package's README file.
     */
    readme?: string;

    /**
     * This package's description.
     */
    description?: string;

    /**
     * An optional link to this package's source code.
     */
    source?: string;

    /**
     * An optional link to this package's issue tracker.
     */
    issues?: string;

    /**
     * An optional link to this package's wiki.
     */
    wiki?: string;

    /**
     * The visibility of a package.
     */
    visibility?: ProjectVisibility;

    /**
     * The license a package is under.
     */
    license?: string;
}

export interface NewPackage {
    /**
     * The name of the package.
     */
    name: string;

    /**
     * The requested slug for the package.
     */
    slug: string;

    /**
     * The package's README file contents.
     */
    readme: string;

    /**
     * A short description of the package.
     */
    description: string;

    /**
     * An optional link to this package's source code.
     */
    source?: string;

    /**
     * An optional link to this package's issue tracker.
     */
    issues?: string;

    /**
     * An optional link to this package's wiki.
     */
    wiki?: string;

    /**
     * The visibility of a package.
     */
    visibility?: ProjectVisibility;

    /**
     * The license a package is under.
     */
    license?: string;
}

export interface PackageData extends Package {
    /**
     * The amount of downloads all of this package's versions have, combined.
     */
    downloads: number;

    /**
     * A list of users that have access to modify this package (a.k.a. authors).
     */
    authors: User[];
}

File: crates/db/src/models/pkg.rs
use super::User;
use crate::schema::packages;
use chrono::NaiveDateTime;
use diesel::pg::Pg;
use diesel_derive_enum::DbEnum;
use itertools::Itertools;

#[derive(
    Debug,
    Clone,
    Copy,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    Hash,
    Serialize,
    Deserialize,
    ToSchema,
    ToResponse,
    DbEnum,
    Default,
)]
#[ExistingTypePath = "crate::schema::sql_types::Visibility"]
pub enum PackageVisibility {
    #[default]
    Public,
    Private,
    Unlisted,
}

/// A package.
#[derive(
    Debug,
    Clone,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    Hash,
    Serialize,
    Deserialize,
    Identifiable,
    Queryable,
    Selectable,
    ToSchema,
    ToResponse,
)]
#[diesel(table_name = packages)]
#[diesel(check_for_backend(Pg))]
pub struct Package {
    /// The package's ID.
    pub id: i32,

    /// The package's name.
    pub name: String,

    /// The package's URL slug.
    pub slug: String,

    /// The package's README.
    pub readme: String,

    /// A short description of the package.
    pub description: String,

    /// The date the package was created.
    pub created_at: NaiveDateTime,

    /// The date the package was last updated.
    pub updated_at: NaiveDateTime,

    /// The amount of downloads a package has.
    pub downloads: i32,

    /// An optional link to the package's source code.
    pub source: Option<String>,

    /// An optional link to the package's issue tracker.
    pub issues: Option<String>,

    /// An optional link to the package's wiki.
    pub wiki: Option<String>,

    /// The visibility of a package.
    pub visibility: PackageVisibility,

    /// The license the package is under.
    pub license: Option<String>,

    /// A list of tags for this package.
    pub tags: Vec<Option<String>>,
}

/// A model for creating a new package.
#[derive(
    Debug,
    Clone,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    Hash,
    Serialize,
    Deserialize,
    Queryable,
    Selectable,
    Insertable,
    ToSchema,
    ToResponse,
)]
#[diesel(table_name = packages)]
#[diesel(check_for_backend(Pg))]
pub struct NewPackage {
    /// The package's URL slug.
    pub slug: String,

    /// The package's name.
    pub name: String,

    /// The package's README.
    pub readme: String,

    /// A short description of the package.
    pub description: String,

    /// An optional link to the package's source code.
    #[serde(default)]
    pub source: Option<String>,

    /// An optional link to the package's issue tracker.
    #[serde(default)]
    pub issues: Option<String>,

    /// An optional link to the package's wiki.
    #[serde(default)]
    pub wiki: Option<String>,

    /// The visibility of a package. Optional. Defaults to public.
    #[serde(default)]
    pub visibility: PackageVisibility,

    /// The license the package is under.
    #[serde(default)]
    pub license: Option<String>,

    /// A list of tags for this package.
    #[serde(default)]
    pub tags: Vec<Option<String>>,
}

/// A package with additional data.
#[derive(
    Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize, ToSchema, ToResponse,
)]
pub struct PackageData {
    /// The package's ID.
    pub id: i32,

    /// The package's name.
    pub name: String,

    /// The package's URL slug.
    pub slug: String,

    /// The package's README.
    pub readme: String,

    /// A short description of the package.
    pub description: String,

    /// An optional link to the package's source code.
    pub source: Option<String>,

    /// An optional link to the package's issue tracker.
    pub issues: Option<String>,

    /// An optional link to the package's wiki.
    pub wiki: Option<String>,

    /// The date the package was created.
    pub created_at: NaiveDateTime,

    /// The date the package was last updated.
    pub updated_at: NaiveDateTime,

    /// The number of downloads the package has.
    pub downloads: i32,

    /// This package's authors.
    pub authors: Vec<User>,

    /// The visibility of a package.
    pub visibility: PackageVisibility,

    /// The license the package is under.
    pub license: Option<String>,

    /// A list of tags for this package.
    pub tags: Vec<String>,
}

impl Package {
    pub fn with_authors(self, authors: Vec<User>) -> PackageData {
        PackageData {
            id: self.id,
            name: self.name,
            slug: self.slug,
            readme: self.readme,
            description: self.description,
            source: self.source,
            issues: self.issues,
            wiki: self.wiki,
            created_at: self.created_at,
            updated_at: self.updated_at,
            downloads: self.downloads,
            visibility: self.visibility,
            license: self.license,
            tags: self.tags.into_iter().filter_map(|v| v).collect_vec(),
            authors,
        }
    }
}

impl PackageVisibility {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Private => "Private",
            Self::Public => "Public",
            Self::Unlisted => "Unlisted",
        }
    }
}
